src/prolog/machine/system_calls.rs

```diff
@@ -1337,6 +1337,68 @@ impl MachineState {
                     }
                 };
             }
+            &SystemClauseType::CharType => {
+                let a1 = self[temp_v!(1)];
+                let a2 = self[temp_v!(2)];
```

Comment:
Try replacing these two lines with:

```
                let a1 = self.store(self.deref(self[temp_v!(1)]));
                let a2 = self.store(self.deref(self[temp_v!(2)]));
```

Reply:
Made the change but the issue is still there.

Reply:
Notice that `'a'` is a member of both symbolic_control_char and alpha_numeric_char. Each call! is independently done. That means there are two unifications of a2 being attempted against two separate heap cell values, which must fail.

Also, if I remember correctly, `';'` isn't a member of any of the character classes in the uncommented call!'s, and there's no `sel.fail = true;` line to make it fail in that case. So, it inertly succeeds without unifying anything to `T`.

Reply:
The code is as if `char_type(a, T) :- T = alnum, T = symbolic_control.` instead of `char_type(a, T) :- T = alnum ; T = symbolic_control.`

Another way to solve this could be:
```Prolog
char_types(a, [alnum, symbolic_control]).
char_type(C, T) :- char_types(C, Ts), member(T, Ts).
```
So the predicate `char_types` would be defined instead of `char_type` directly but a list of atoms needs to be constructed.